File: array/merge-sort.js
// Implement merge sort: "divide and conquer" algorithm using recursion

function mergeSort(arr) {
	if (arr.length < 2) {
		return arr;
	}

	var middle = parseInt(arr.length / 2),
		left = arr.slice(0, middle),
		right = arr.slice(middle, arr.length);

	return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
	var result = [];

	while (left.length && right.length) {
		if (left[0] <= right[0]) {
			result.push(left.shift());
		} else {
			result.push(right.shift());
		}
	}

	while (left.length) {
		result.push(left.shift());
	}

	while (right.length) {
		result.push(right.shift());
	}

	return result;
}

// test case
mergeSort([5, 4, 3, 2, 1]);

File: array/merge-sort.py
# python implementation of merge sort

def merge_sort(arr):
	if len(arr) > 1:
		middle = len(arr)/2
		left = arr[:middle]
		right = arr[middle:]

		merge_sort(left)
		merge_sort(right)

		i = 0
		j = 0
		k = 0

		while i < len(left) and j < len(right):
			if left[i] < right[j]:
				arr[k] = left[i]
				i = i + 1
			else:
				arr[k] = right[j]
				j = j + 1
			k = k + 1

		while i < len(left):
			arr[k] = left[i]
			i = i + 1
			k = k + 1

		while j < len(right):
			arr[k] = right[j]
			j = j + 1
			k = k + 1

	return arr


# test case
print merge_sort([5, 4, 3, 2, 1])
